pkg/providers/ibmcloud.go

```diff
@@ -0,0 +1,317 @@
+		if itemPrice.LocationGroupId == nil {
+			return itemPrice
+		}
+	}
+	return datatypes.Product_Item_Price{}
+
+}
+
+func (p *ibmcloudProvider) getItemPriceList(packageId int, itemKeyNames []string) (resp []datatypes.Product_Item_Price) {
+
+	items := p.getPackageItems(packageId)
+	var prices []datatypes.Product_Item_Price
+
+	for _, itemKeyName := range itemKeyNames {
+		for _, item := range items {
+			if (*item.KeyName) == itemKeyName {
+				itemPrice := p.getStandardPrice(item)
+				prices = append(prices, itemPrice)
+				break
+			}
+		}
+	}
+
+	return prices
+}
+
+func (p *ibmcloudProvider) serverCreate(pool, presetname string) (string, error) {
+
+	location := locations[rand.Intn(len(locations))]
+
+	hostname := strings.Split(uuid.New().String(), "-")[0]
+	domain := pool + ".ofcir"
+
+	key, err := p.ensureSSHKey(p.config.Sshkey)
+	if err != nil {
+		return "", err
+	}
+
+	server := []datatypes.Hardware{
+		{
+			Hostname: sl.String(hostname),
+			Domain:   sl.String(domain),
+		},
+	}
+
+	// IBMC Servers have to be odered with a range of packages, this is a list of the
+	// minimum required by hourly BM nodes e.g. 1_IP_ADDRESS vs 4_PUBLIC_IP_ADDRESSES
+	// Get a list of available keys with
+	// ibmcloud sl order item-list  BARE_METAL_SERVER
+	required_items := []string{
+		"REBOOT_KVM_OVER_IP",
+		"UNLIMITED_SSL_VPN_USERS_1_PPTP_VPN_USER_PER_ACCOUNT",
+		"OS_CENTOS_STREAM_8_X_64_BIT",
+		"BANDWIDTH_0_GB_2",
+		"1_GBPS_PUBLIC_PRIVATE_NETWORK_UPLINKS",
+		"1_IP_ADDRESS",
+	}
+	// Build a skeleton SoftLayer_Product_Item_Price objects.
+	prices := p.getItemPriceList(IbmCloudFastProvisioningServer, required_items)
+
+	// We need to find the ID of the preset with the required name
+	packageService := services.GetProductPackageService(p.client)
+
+	// ibmcloud sl order preset-list  --prices BARE_METAL_SERVER
+	// e.g. presetname == 1U_4210S_384GB_2X4TB_RAID_1
+	// 1U_4210S_384GB_2X4TB_RAID_1, id = 1278
+	presets, err := packageService.Id(IbmCloudFastProvisioningServer).GetActivePresets()
+	presetId := -1
+	for _, preset := range presets {
+		if *preset.KeyName == presetname {
+			presetId = *preset.Id
+			break
+		}
+	}
+	if presetId == -1 {
+		return "", errors.New("Can't find preset with name " + presetname)
+	}
+
+	orderkeys := []datatypes.Container_Product_Order_SshKeys{
+		{
+			SshKeyIds: []int{*key},
+		},
+	}
+
+	// Build a container_Product_Order object.
+	orderTemplate := datatypes.Container_Product_Order{
+		Quantity:         sl.Int(1),
+		Location:         sl.String(location),
+		PackageId:        sl.Int(IbmCloudFastProvisioningServer),
+		PresetId:         sl.Int(presetId),
+		Prices:           prices,
+		Hardware:         server,
+		UseHourlyPricing: sl.Bool(true),
+		SshKeys:          orderkeys,
+	}
+
+	// Get SoftLayer_Product_Order service.
+	service := services.GetProductOrderService(p.client)
+
+	// Uncomment to "dry-run" the order before placing it
+	//result, err := service.VerifyOrder(&orderTemplate)
```

Comment:
Is this required only for testing or do you plan something else?

Reply:
Its only there for testing, I wanted to just leave it there as a reminder that its possible (if testing changes in future)